akka-projection-cassandra/src/main/scala/akka/projection/cassandra/internal/CassandraProjectionImpl.scala

```diff
@@ -73,115 +72,141 @@ import akka.stream.scaladsl.Source
+          case (offset, envelope) =>
+            applyUserRecovery(handler, envelope, offset, logger, () => handler.process(envelope))
+              .map(_ => offset)
         }
-      case AtLeastOnce(afterEnvelopes, orAfterDuration) =>
-        source
-          .via(handlerFlow)
-          .groupedWithin(afterEnvelopes, orAfterDuration)
-          .collect { case grouped if grouped.nonEmpty => grouped.last }
-          .mapAsync(parallelism = 1) { offset =>
+
+      val composedSource: Source[Done, NotUsed] = strategy match {
+        case AtLeastOnce(1, _) =>
+          // optimization of general AtLeastOnce case
+          source.via(handlerFlow).mapAsync(1) { offset =>
             offsetStore.saveOffset(projectionId, offset)
           }
-      case AtMostOnce =>
-        // prevent usage of retries for at-most-once
-        val atMostOnceHandler = new AtMostOnceHandler(handler, logger)
-        source
-          .mapAsync(parallelism = 1) {
-            case (offset, envelope) =>
-              offsetStore
-                .saveOffset(projectionId, offset)
-                .flatMap(
-                  _ =>
-                    applyUserRecovery(
-                      atMostOnceHandler,
-                      envelope,
-                      offset,
-                      logger,
-                      () => atMostOnceHandler.process(envelope)))
-          }
-          .map(_ => Done)
+        case AtLeastOnce(afterEnvelopes, orAfterDuration) =>
+          source
+            .via(handlerFlow)
+            .groupedWithin(afterEnvelopes, orAfterDuration)
+            .collect { case grouped if grouped.nonEmpty => grouped.last }
+            .mapAsync(parallelism = 1) { offset =>
+              offsetStore.saveOffset(projectionId, offset)
+            }
+        case AtMostOnce =>
+          // prevent usage of retries for at-most-once
+          val atMostOnceHandler = new AtMostOnceHandler(handler, logger)
+          source
+            .mapAsync(parallelism = 1) {
+              case (offset, envelope) =>
+                offsetStore
+                  .saveOffset(projectionId, offset)
+                  .flatMap(
+                    _ =>
+                      applyUserRecovery(
+                        atMostOnceHandler,
+                        envelope,
+                        offset,
+                        logger,
+                        () => atMostOnceHandler.process(envelope)))
+            }
+            .map(_ => Done)
+      }
+
+      composedSource
+    }
+
+    private[projection] def newRunningInstance(): RunningProjection = {
+      new CassandraRunningProjection(RunningProjection.withBackoff(mappedSource(), settings), killSwitch)
     }
+  }
 
-    composedSource
+  private class CassandraRunningProjection(source: Source[Done, _], killSwitch: SharedKillSwitch)(
+      implicit val systemProvider: ClassicActorSystemProvider)
+      extends RunningProjection {
+
+    private val futureDone = source.run()
```

Comment:
This is nice simplification. Not AtmoicBoolean, no Promises, just a Future[Done]